Source/ASDisplayTree.h

```diff
@@ -0,0 +1,112 @@
+//
+//  ASDisplayTree.h
+//  AsyncDisplayKit
+//
+//  Created by Adlai on 5/28/18.
+//  Copyright © 2018 Pinterest. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import <AsyncDisplayKit/ASBaseDefines.h>
+#import <AsyncDisplayKit/ASTraitCollection.h>
+
+@class ASDisplayNode;
+
+typedef NS_OPTIONS(NSUInteger, ASTreeEnumerationOptions) {
+  ASEnumerateBreadthFirst  = 1 << 0,  // Default is depth-first.
+  ASEnumerateSkipSelf   = 1 << 1,     // Default is to include self.
+};
+
+typedef NS_ENUM(NSInteger, ASTreeInsertionLocation) {
+  ASTreeInsertBelow,      // Other node is sibling above. Index ignored.
+  ASTreeInsertReplace,    // Other node is replaced. Index ignored.
+  ASTreeInsertAbove,      // Other node is sibling below. Index ignored.
+  ASTreeInsertAtEnd,      // Other node is parent. Index ignored.
+  ASTreeInsertAtIndex     // Other node is parent.
+};
+
+NS_ASSUME_NONNULL_BEGIN
+
+/*
+ * An entire tree of display nodes. Subtrees are not
```

Comment:
Maybe better: "An entire tree of display nodes, always owned by a root node that doesn't have a .supernode"

Wrapping is pretty aggressive here, might make sense to use the standard ~120 column line length for Obj-C.